src/System.Memory/src/System/Buffers/ReadOnlySequence.cs

```diff
@@ -276,7 +278,23 @@ public ReadOnlySequence<T> Slice(long start)
         }
 
         /// <inheritdoc />
-        public override string ToString() => string.Format("System.Buffers.ReadOnlySequence<{0}>[{1}]", typeof(T).Name, Length);
+        public override string ToString()
+        {
+            if (typeof(T) == typeof(char) && Length < int.MaxValue)
+            {
+                return string.Create((int)Length, this, (span, sequence) =>
+                {
+                    foreach (ReadOnlyMemory<T> readOnlyMemory in sequence)
+                    {
+                        var sourceSpan = ((ReadOnlyMemory<char>)(object)readOnlyMemory).Span;
```

Comment:
There should be a better way of doing this.

Reply:
`Unsafe.As` is another option; but just as ugly

Reply:
nit: state type explicitly as `ReadOnlySpan<char>`, especially since there is type conversions happening here

> There should be a better way of doing this.

cc @VSadov, @AndyAyersMS

Reply:
`Unsafe.As ` requires reference type.

Reply:
`ReadOnlyMemory<T>` and `ReadOnlyMemory<char>` are basically unrelated types in terms of conversions. 
Converting to some intermediate type (generally object) and then again is how such things are converted. This is not specific to `ReadOnlyMemory`

Reply:
Something like this? Though, as @benaadams mentioned, it isn't pretty.
```C#
ReadOnlySpan<char> span = Unsafe.As<ReadOnlyMemory<T>, ReadOnlyMemory<char>>(ref readOnlyMemory).Span;
```

Reply:
Otherwise a safe conversion needs a typecheck on the instance and structs know their type only when in a boxed form.
You cant have this both formally safe and without boxing.